### ISSUE TITLE
Possible incompatibility with Airbrake gem

### ISSUE THREAD

**Issue by raivil** [googleapis/google-cloud-ruby#2492]
Hi,

I'm upgrading a Rails application to version 5.2.1 and some gems were updated.
```
ruby 2.4.2
airbrake-7.3.5
airbrake-ruby-2.11.0
google-cloud-logging-1.5.2
```

Looks like all executables (rails console, db migrations), when exiting, are raising an exception. This causes the return value from executables to be 1.

I'm unsure if this is an airbrake issue or a gcloud logging issue.

Can anyone help?

Example:
```
# rails c
Loading staging environment (Rails 5.2.1)
irb(main):002:0> Rails.logger.info("test")
test
=> true
irb(main):003:0> Rails.logger
=> #<Google::Cloud::Logging::Logger:0x000055e6b3c77438 @writer=#<Google::Cloud::Logging::AsyncWriter:0x000055e6b3c77640 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x000055e6b3c77578>, @startup_lock=#<Thread::Mutex:0x000055e6b3c77550>, @cleanup_options={:wait_interval=>1.0, :timeout=>10.0, :force=>true}, @logging=#<Google::Cloud::Logging::Project:0x000055e6b3c651e8 @service=Google::Cloud::Logging::Service(projectname), @shared_async_writer=#<Google::Cloud::Logging::AsyncWriter:0x000055e6b3c77640 ...>>, @max_queue_size=10000, @partial_success=false, @queue_resource=#<MonitorMixin::ConditionVariable:0x000055e6b3c774d8 @monitor=#<Google::Cloud::Logging::AsyncWriter:0x000055e6b3c77640 ...>, @cond=#<Thread::ConditionVariable:0x000055e6b3c774b0>>, @queue=[], @queue_size=0, @lock_cond=#<MonitorMixin::ConditionVariable:0x000055e6b790b660 @monitor=#<Google::Cloud::Logging::AsyncWriter:0x000055e6b3c77640 ...>, @cond=#<Thread::ConditionVariable:0x000055e6b790b638>>, @async_state=:running, @thread=#<Thread:0x000055e6b790b2f0@/usr/local/bundle/gems/stackdriver-core-1.3.0/lib/stackdriver/core/async_actor.rb:314 sleep_forever>, @last_exception=#<Google::Cloud::PermissionDeniedError: 7:The caller does not have permission>>, @log_name="ruby_app_log", @resource=#<Google::Cloud::Logging::Resource:0x000055e6b3c77690 @labels={:cluster_name=>"environment", :namespace_id=>"default"}, @type="container">, @labels={}, @level=0, @request_info={}, @closed=false, @formatter=#<Logger::Formatter:0x000055e6b3c773c0 @datetime_format=nil>, @datetime_format="", @silencer=true, @project="projectname">
irb(main):004:0> 
**Airbrake: closed
/usr/local/bundle/gems/google-cloud-logging-1.5.2/lib/google/cloud/logging/async_writer.rb:155:in `block in write_entries': AsyncWriter has been stopped (RuntimeError)
	from /usr/local/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.2/lib/google/cloud/logging/async_writer.rb:154:in `write_entries'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.2/lib/google/cloud/logging/logger.rb:553:in `write_entry'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.2/lib/google/cloud/logging/logger.rb:317:in `add'
	from /usr/local/bundle/gems/activesupport-5.2.1/lib/active_support/logger.rb:28:in `block (2 levels) in broadcast'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.2/lib/google/cloud/logging/logger.rb:195:in `debug'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby/async_sender.rb:53:in `close'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby/notifier.rb:73:in `close'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby.rb:280:in `close'
	from /usr/local/bundle/gems/airbrake-7.3.5/lib/airbrake.rb:29:in `block in <main>'

# echo $?
1

```

**Comment by raivil** [comment]
@blowmage The app doesn't have a lof of specific configs and nothing specific (or custom) to stop those.

I can't reproduce this error on macos, only on a docker container. 
I'm still investigating and testing configurations to see if anything changes.

Config samples:
Airbrake config is pretty much default
```
....
  c.logger = Rails.logger
  c.environment = Rails.env
  c.ignore_environments = %w(development test)
  c.blacklist_keys = Rails.application.config.filter_parameters

Airbrake.add_filter do |notice|
  notice.ignore! if notice[:errors].any? do |error|
    %w(
      Sidekiq::Shutdown and others.
)
end
require "airbrake/sidekiq"
```

and Stackdriver code is super simple:
```
# StackDrive configs (on environment files)
  config.google_cloud.use_error_reporting = true
  config.google_cloud.use_debugger = true
  config.google_cloud.use_logging = true
  config.google_cloud.use_trace = true
  config.google_cloud.logging.log_name_map =  { .... }
  ```

**Comment by blowmage** [comment]
> I can't reproduce this error on macos, only on a docker container.

Ah, very interesting. Can you provide more information? Host platform? Container OS? Base image? etc.

**Comment by raivil** [comment]
GKE 1.9.7-gke.6 with Container-Optimized OS (cos)
ruby official image: ruby:2.4.2-stretch

**Comment by kyrylo** [comment]
Airbrake Ruby developer here.

The race assumption is interesting and seems to be a good clue. To support that theory, I can add that Airbrake Ruby doesn't attempt to send any exceptions when it's closed. We have a check against that:

https://github.com/airbrake/airbrake-ruby/blob/64e91179f5a25f2fe35e3deb68e36e84e4ef98bc/lib/airbrake-ruby/notifier.rb#L58-L61

I am not familiar with `google-cloud-logging`, but I would guess this check makes the same thing:

https://github.com/googleapis/google-cloud-ruby/blob/135de4ae938201789d1c2c56c8f0d573b9f22c4d/google-cloud-logging/lib/google/cloud/logging/async_writer.rb#L155

I have nothing to suggest yet, but I am following this issue.

**Comment by blowmage** [comment]
@raivil do you only see this behavior when being run from a container? Or when being run in production? Are you using a forking web server? If so, which one?

**Comment by raivil** [comment]
@blowmage I see this error only on a container in staging. Happens running rake db:migrate or after_party:run, rails console, sidekiq, but not puma (v3.8.2 with one worker).

**Comment by blowmage** [comment]
Can you try reproducing in a separate dummy app that only adds Google Cloud Logging and Airbrake, and any other dependency that is needed to demonstrate the behavior? Having a known-working reproducible app would help the diagnosis immensely.

**Comment by blowmage** [comment]
@raivil Can you try upgrading to [1.5.6](https://github.com/googleapis/google-cloud-ruby/releases/tag/google-cloud-logging%2Fv1.5.6) and let us know if you still see this behavior?

**Comment by raivil** [comment]
@blowmage I'm not able to reproduce the error locally with the app or another new app.
I'll try the new version and let you know.

**Comment by raivil** [comment]
Same issue with 1.5.6.
```
Executing: RAILS_ENV=staging  exec bundle exec rake db:migrate
/usr/local/bundle/gems/google-cloud-logging-1.5.6/lib/google/cloud/logging/async_writer.rb:155:in `block in write_entries': AsyncWriter has been stopped (RuntimeError)
	from /usr/local/lib/ruby/2.4.0/monitor.rb:214:in `mon_synchronize'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.6/lib/google/cloud/logging/async_writer.rb:154:in `write_entries'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.6/lib/google/cloud/logging/logger.rb:555:in `write_entry'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.6/lib/google/cloud/logging/logger.rb:326:in `add'
	from /usr/local/bundle/gems/google-cloud-logging-1.5.6/lib/google/cloud/logging/logger.rb:202:in `debug'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby/async_sender.rb:53:in `close'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby/notifier.rb:73:in `close'
	from /usr/local/bundle/gems/airbrake-ruby-2.11.0/lib/airbrake-ruby.rb:280:in `close'
	from /usr/local/bundle/gems/airbrake-7.3.5/lib/airbrake.rb:29:in `block in <main>'
```

**Comment by blowmage** [comment]
Thanks for trying. I am still not seeing where or how the Logger is being closed. Airbrake logs that it is being closed, and the error is being raised because the Logger is already closed. I’ll keep looking. Any reproduction of this behavior that you can share is very much appreciated.

**Comment by blowmage** [comment]
@raivil I was looking through your previous comment and noticed something I missed earlier. The inspect of the `Rails.logger` object shows that is is a `Google::Cloud::Logging::Logger` object, and it also shows the logger's AsyncWriter has this:

```
@last_exception=#<Google::Cloud::PermissionDeniedError: 7:The caller does not have permission>
```

This ivar is set on the writer object when there is an error in the background thread. The background thread is where the API call to the Logging service is made. That is most likely where the writer is stopped.

So, I would assume that solving the permission error (most likely an authentication problem) will clear this up. My apologies for not seeing this sooner.

**Comment by raivil** [comment]
@blowmage I'll dig further on this. i'm not sure what permisison may be missing because logs are being sent to stackdriver as expected.

Also I just realized that before upgrading rails, while using 4.2.x , the `Rails.loger` is `RailsStdoutLogging::StdoutLogger:`. 
On Rails 5.2 with the "same" config the `Rails.logger` is a `Google::Cloud::Logging::Logger`. 
One big difference is that we removed rails_12factor gem because it's already included in rails.

**Comment by blowmage** [comment]
@raivil Have you been able to confirm whether or not your staging environment is configured for Google Cloud correctly?

**Comment by raivil** [comment]
@blowmage not yet.

**Comment by blowmage** [comment]
@raivil Have you been able to confirm your staging environment configuration?

**Comment by raivil** [comment]
@blowmage 
I'm still trying to figure it out.

The code snipped that you sent has an error as the `@writer` is not an `Google::Cloud::Logging::AsyncWriter`

```
irb(main):001:0> old_logger = Rails.logger
=> #<Google::Cloud::Logging::Logger:0x0000556345a144d0 @writer=#<Google::Cloud::Logging::AsyncWriter:0x0000556345a148e0 @mon_owner=nil, @mon_count=0, @mon_mutex=#<Thread::Mutex:0x0000556345a146d8>, @startup_lock=#<Thread::Mutex:0x0000556345a146b0>, @cleanup_options={:wait_interval=>1.0, :timeout=>10.0, :force=>true}, @logging=#<Google::Cloud::Logging::Project:0x0000556345a1baa0 @service=Google::Cloud::Logging::Service(project-account), @shared_async_writer=#<Google::Cloud::Logging::AsyncWriter:0x0000556345a148e0 ...>>, @max_queue_size=10000, @partial_success=false, @queue_resource=#<MonitorMixin::ConditionVariable:0x0000556345a145e8 @monitor=#<Google::Cloud::Logging::AsyncWriter:0x0000556345a148e0 ...>, @cond=#<Thread::ConditionVariable:0x0000556345a145c0>>, @queue=[], @queue_size=0>, @log_name="ruby_app_log", @resource=#<Google::Cloud::Logging::Resource:0x0000556345a14930 @labels={:cluster_name=>"staging", :namespace_id=>"default"}, @type="container">, @labels={}, @level=0, @request_info_var=#<Concurrent::ThreadLocalVar:0x0000556345a14480 @default_block=nil, @default=nil, @index=0>, @closed=false, @formatter=#<Logger::Formatter:0x0000556345a14368 @datetime_format=nil>, @datetime_format="", @silencer=true, @project="project-account">
irb(main):002:0> raise unless old_logger.is_a? Google::Cloud::Logging::Logger
=> nil
irb(main):003:0> Rails.logger = Google::Cloud::Logging::Logger.new(
irb(main):004:1*   old_logger.project, old_logger.log_name, old_logger.resource, old_logger.labels
irb(main):005:1> )
=> #<Google::Cloud::Logging::Logger:0x000055634a4ead10 @writer="project-account", @log_name="ruby_app_log", @resource=#<Google::Cloud::Logging::Resource:0x0000556345a14930 @labels={:cluster_name=>"staging", :namespace_id=>"default"}, @type="container">, @labels={}, @level=0, @request_info_var=#<Concurrent::ThreadLocalVar:0x000055634a4eac48 @default_block=nil, @default=nil, @index=2>, @closed=false, @formatter=#<Logger::Formatter:0x000055634a4eab58 @datetime_format=nil>, @datetime_format="", @silencer=true>
irb(main):006:0> Rails.logger.debug("test from staging environment") # Makes API call
NoMethodError: undefined method `write_entries' for "project-account":String
  from (irb):6
irb(main):007:0> 
```

**Comment by blowmage** [comment]
Oops, I had a mistake in my code. The first argument should be this instead:

```ruby
old_logger.writer.logging
```

So the code should really be this:

```ruby
old_logger = Rails.logger
raise unless old_logger.is_a? Google::Cloud::Logging::Logger
Rails.logger = Google::Cloud::Logging::Logger.new(
  old_logger.writer.logging, old_logger.log_name, old_logger.resource, old_logger.labels
)
Rails.logger.debug("test from staging environment") # Makes API call
```

**Comment by raivil** [comment]
@blowmage Thanks.
Executing that I get a `Google::Cloud::PermissionDeniedError: 7:The caller does not have permission` as mentioned before.

I'm working with my Infrastructure team to understand this permisison issue.
I believe nothing was changed regarding permissions on our staging accounts, but we'll check it out.
The problem appeared right after deploying the rails 5 update.

Also staging environemnt is completely isolated from production, so we're note reusing any credentials/acocunts/permissions.

**Comment by blowmage** [comment]
Does calling `Rails.logger.writer.logging.warn("another test")` also work?

**Comment by raivil** [comment]
Private method there, but calling `Rails.logger.writer.logging.send(:warn,"another test")` works.

```
....
 severity:  "WARNING"  
 textPayload:  "another test"  
 timestamp:  "2018-10-17T20:05:50.624103090Z"  
```

**Comment by blowmage** [comment]
My apologies again, I gave you more bad code to try. I'll try to get my head in the game.

I'm trying to make it as simple as possible to reuse the configuration to make an API call. I think the simplest is to run the code I posted previously:

```
$ rails c
Loading staging environment (Rails 5.2.1)
irb(main):001:0> old_logger = Rails.logger
irb(main):002:0> raise unless old_logger.is_a? Google::Cloud::Logging::Logger
irb(main):003:0> Rails.logger = Google::Cloud::Logging::Logger.new(old_logger.writer.logging, old_logger.log_name, old_logger.resource, old_logger.labels)
irb(main):004:0> Rails.logger.warn("make api call directly")
```

Does that work? The code you tried previously is not calling Google Cloud Logger API, and is the equivalent to doing this:

```
$ irb
irb(main):001:0> obj = Object.new
=> #<Object:0x00007fb2648d3218>
irb(main):002:0> obj.send :warn, "hi"
hi
=> nil
```

**Comment by raivil** [comment]
@blowmage Yes, that code works and I can see the logs on stackdriver.

**Comment by blowmage** [comment]
Okay, verifying the error still occurs when staging environment is properly configured is very helpful. I'll try to reproduce this again. Any other information about your application that you can share is very helpful. A clean dummy app that reproduces this is the most helpful, of course.

The error is happening because the logger object is closed (either by an error or manually) when Airbrake logs messages as the process exits. I don't know what is doing this to the the logger.

**Comment by blowmage** [comment]
@raivil Did you mean to close this issue? Did you find the cause?

**Comment by raivil** [comment]
@blowmage Github closed it automatically based on another private pr of mine.

**Comment by danielnc** [comment]
Any updates? 🆙

**Comment by blowmage** [comment]
We have changes coming to AsyncWriter that will _not_ close the writer object when it gets an error. I suspect this change will fix this issue. But until I can reproduce I can't be 100% confident.

**Comment by blowmage** [comment]
The changes mentioned previously have (finally) made their way to master, but we are in a release freeze over the holidays so we won't see a release until next year. However, you can try these changes out now if you add the following to your `Gemfile`:

```ruby
gem "google-cloud-logging",  github: "googleapis/google-cloud-ruby"
```

@danielnc Can you try this and let me know if these changes address your problem? You may also consider setting `Google::Cloud::Logging.configure.on_error` to a lambda that will report errors. Errors due to authentication issues should show up there.

**Comment by blowmage** [comment]
@raivil Can you try using the code in master to see if it resolves this behavior for you?

**Comment by blowmage** [comment]
Closing to lack of response.

We have made several changes to the stackdriver gems address running in multi-threaded webservers such as puma and unicorn. And we have added the `#on_error` callback that will be called when errors are raised on background threads, and made background actors more resilient to errors. The only way I could reproduce the behavior in this issue was to raise such an error. So, I am pretty confident that the underlying problem for this issue has been addressed. We just haven't gotten verification.
